File: swift/Sources/EnvdoctorKit/Config/TomlParser.swift
import Foundation

/// A minimal TOML parser covering the subset envdoctor configs use: string /
/// integer / float / boolean values, arrays, inline tables, `[table]`
/// sections, dotted keys, and comments. Produces nested
/// `[String: Any?]`-shaped values mirroring Tomlyn's `TomlTable` model.
public enum TomlParser {
    public struct Error: Swift.Error, CustomStringConvertible {
        public let description: String
    }

    /// Reference-typed table so nested `[section]` headers can be navigated
    /// and mutated without fighting Swift dictionary value semantics.
    private final class Table {
        var entries: [String: Any?] = [:]
    }

    public static func parse(_ content: String) throws -> [String: Any?] {
        let root = Table()
        var current = root

        let lines = content.components(separatedBy: "\n")
        for (idx, rawLine) in lines.enumerated() {
            let line = stripComment(rawLine).trimmingCharacters(in: .whitespaces)
            if line.isEmpty { continue }
            let lineNo = idx + 1

            if line.hasPrefix("[") {
                guard line.hasSuffix("]") else {
                    throw Error(description: "line \(lineNo): malformed table header")
                }
                let inner = String(line.dropFirst().dropLast()).trimmingCharacters(in: .whitespaces)
                let path = try parseKeyPath(inner, lineNo)
                var node = root
                for part in path {
                    if let existing = node.entries[part] as? Table {
                        node = existing
                    } else {
                        let table = Table()
                        node.entries[part] = table
                        node = table
                    }
                }
                current = node
                continue
            }

            guard let eq = findEqualsOutsideString(line) else {
                throw Error(description: "line \(lineNo): expected key = value")
            }
            let keyPath = try parseKeyPath(String(line[..<eq]).trimmingCharacters(in: .whitespaces), lineNo)
            let value = try parseValue(String(line[line.index(after: eq)...]).trimmingCharacters(in: .whitespaces), lineNo)
            try assign(&current.entries, keyPath: keyPath, value: value, lineNo: lineNo)
        }
        return plain(root)
    }

    private static func plain(_ table: Table) -> [String: Any?] {
        var out: [String: Any?] = [:]
        for (key, value) in table.entries {
            switch value {
            case let child as Table:
                out[key] = plain(child)
            case let list as [Any?]:
                out[key] = list.map { item -> Any? in
                    if let t = item as? Table { return plain(t) }
                    if let l = item as? [Any?] { return l }
                    return item
                }
            default:
                out[key] = value
            }
        }
        return out
    }

    private static func assign(_ entries: inout [String: Any?], keyPath: [String], value: Any?, lineNo: Int) throws {
        if keyPath.count == 1 {
            if entries[keyPath[0]] != nil {
                throw Error(description: "line \(lineNo): duplicate key '\(keyPath[0])'")
            }
            entries[keyPath[0]] = value
            return
        }
        let head = keyPath[0]
        if entries[head] == nil {
            entries[head] = Table()
        }
        guard let table = entries[head] as? Table else {
            throw Error(description: "line \(lineNo): key '\(head)' is not a table")
        }
        try assign(&table.entries, keyPath: Array(keyPath[1...]), value: value, lineNo: lineNo)
    }

    private static func stripComment(_ line: String) -> String {
        var inString = false
        var quoteChar: Character = " "
        var escaped = false
        var result = ""
        for ch in line {
            if inString {
                result.append(ch)
                if escaped {
                    escaped = false
                } else if ch == "\\" && quoteChar == "\"" {
                    escaped = true
                } else if ch == quoteChar {
                    inString = false
                }
            } else if ch == "\"" || ch == "'" {
                inString = true
                quoteChar = ch
                result.append(ch)
            } else if ch == "#" {
                break
            } else {
                result.append(ch)
            }
        }
        return result
    }

    private static func findEqualsOutsideString(_ line: String) -> String.Index? {
        var inString = false
        var quoteChar: Character = " "
        var index = line.startIndex
        while index < line.endIndex {
            let ch = line[index]
            if inString {
                if ch == quoteChar { inString = false }
            } else if ch == "\"" || ch == "'" {
                inString = true
                quoteChar = ch
            } else if ch == "=" {
                let prev = index > line.startIndex ? line[line.index(before: index)] : " "
                let next = line.index(after: index)
                let prevIsOp = prev == "=" || prev == "<" || prev == ">" || prev == "!"
                let nextIsEq = next < line.endIndex && line[next] == "="
                if !prevIsOp && !nextIsEq {
                    return index
                }
            }
            index = line.index(after: index)
        }
        return nil
    }

    private static func parseKeyPath(_ text: String, _ lineNo: Int) throws -> [String] {
        try text.split(separator: ".").map { part in
            let trimmed = part.trimmingCharacters(in: .whitespaces)
            if trimmed.hasPrefix("\"") || trimmed.hasPrefix("'") {
                guard trimmed.count >= 2 else { throw Error(description: "line \(lineNo): bad quoted key") }
                return String(trimmed.dropFirst().dropLast())
            }
            guard !trimmed.isEmpty else { throw Error(description: "line \(lineNo): empty key") }
            return trimmed
        }
    }

    private static func parseValue(_ text: String, _ lineNo: Int) throws -> Any? {
        let value = text.trimmingCharacters(in: .whitespaces)
        if value.isEmpty { throw Error(description: "line \(lineNo): missing value") }
        if value.hasPrefix("\"") {
            return try parseBasicString(value, lineNo)
        }
        if value.hasPrefix("'") {
            guard value.hasSuffix("'"), value.count >= 2 else {
                throw Error(description: "line \(lineNo): unterminated literal string")
            }
            return String(value.dropFirst().dropLast())
        }
        if value.hasPrefix("[") {
            return try parseArray(value, lineNo)
        }
        if value.hasPrefix("{") {
            return try parseInlineTable(value, lineNo)
        }
        switch value {
        case "true": return true
        case "false": return false
        default: break
        }
        // TOML requires underscores may separate digits; normalize them.
        let numeric = value.replacingOccurrences(of: "_", with: "")
        if let i = Int(numeric) { return i }
        if let d = Double(numeric) { return d }
        // Datetimes and bare tokens are outside the config subset.
        throw Error(description: "line \(lineNo): invalid value '\(value)'")
    }

    private static func parseBasicString(_ value: String, _ lineNo: Int) throws -> String {
        var result = ""
        var i = value.index(after: value.startIndex)
        while i < value.endIndex {
            let ch = value[i]
            if ch == "\"" {
                return result
            }
            if ch == "\\" {
                let next = value.index(after: i)
                guard next < value.endIndex else { break }
                let esc = value[next]
                switch esc {
                case "n": result.append("\n")
                case "t": result.append("\t")
                case "r": result.append("\r")
                case "\"": result.append("\"")
                case "\\": result.append("\\")
                case "b": result.append("\u{08}")
                case "f": result.append("\u{0C}")
                case "u":
                    let hexStart = value.index(after: next)
                    let hexEnd = value.index(hexStart, offsetBy: 4, limitedBy: value.endIndex) ?? value.endIndex
                    guard let code = UInt32(String(value[hexStart..<hexEnd]), radix: 16),
                          let scalar = Unicode.Scalar(code) else {
                        throw Error(description: "line \(lineNo): bad \\u escape")
                    }
                    result.unicodeScalars.append(scalar)
                    i = hexEnd
                    continue
                default:
                    throw Error(description: "line \(lineNo): bad escape '\\\(esc)'")
                }
                i = value.index(after: next)
                continue
            }
            result.append(ch)
            i = value.index(after: i)
        }
        throw Error(description: "line \(lineNo): unterminated string")
    }

    private static func splitTopLevel(_ text: String) -> [String] {
        var parts: [String] = []
        var depth = 0
        var inString = false
        var quoteChar: Character = " "
        var current = ""
        for ch in text {
            if inString {
                current.append(ch)
                if ch == quoteChar { inString = false }
                continue
            }
            if ch == "\"" || ch == "'" {
                inString = true
                quoteChar = ch
                current.append(ch)
            } else if ch == "[" || ch == "{" {
                depth += 1
                current.append(ch)
            } else if ch == "]" || ch == "}" {
                depth -= 1
                current.append(ch)
            } else if ch == "," && depth == 0 {
                parts.append(current)
                current = ""
            } else {
                current.append(ch)
            }
        }
        if !current.trimmingCharacters(in: .whitespaces).isEmpty {
            parts.append(current)
        }
        return parts
    }

    private static func parseArray(_ value: String, _ lineNo: Int) throws -> [Any?] {
        guard value.hasSuffix("]") else {
            throw Error(description: "line \(lineNo): unterminated array")
        }
        let inner = String(value.dropFirst().dropLast())
        return try splitTopLevel(inner).map { try parseValue($0, lineNo) }
    }

    private static func parseInlineTable(_ value: String, _ lineNo: Int) throws -> [String: Any?] {
        guard value.hasSuffix("}") else {
            throw Error(description: "line \(lineNo): unterminated inline table")
        }
        let inner = String(value.dropFirst().dropLast())
        let table = Table()
        for part in splitTopLevel(inner) {
            let trimmed = part.trimmingCharacters(in: .whitespaces)
            guard let eq = findEqualsOutsideString(trimmed) else {
                throw Error(description: "line \(lineNo): expected key = value in inline table")
            }
            let keyPath = try parseKeyPath(String(trimmed[..<eq]).trimmingCharacters(in: .whitespaces), lineNo)
            let raw = String(trimmed[trimmed.index(after: eq)...]).trimmingCharacters(in: .whitespaces)
            try assign(&table.entries, keyPath: keyPath, value: try parseValue(raw, lineNo), lineNo: lineNo)
        }
        return table.entries
    }
}

File: kotlin/src/main/kotlin/io/github/arun_skg/envdoctor/utils/TomlParser.kt
package io.github.arun_skg.envdoctor.utils

/**
 * A small hand-rolled TOML parser covering the subset envdoctor's config
 * uses: comments, bare/quoted keys, `key = value` pairs with basic strings
 * (single-line, escapes), integers, booleans, arrays (possibly spanning
 * lines, possibly of inline tables), inline tables, and standard table
 * headers (`[table]`, `[table.sub]`).
 *
 * Produces plain values: `JsonObject` (LinkedHashMap), `List<Any?>`, `String`,
 * `Boolean`, and numbers as `Long`.
 */
object TomlParser {
    fun parse(text: String): JsonObject = Parser(text).parse()

    private class Parser(val text: String) {
        private var i = 0
        private val n = text.length

        private val root = JsonObject()
        private var current: JsonObject = root

        private fun error(msg: String): Nothing = throw IllegalArgumentException("TOML: $msg near offset $i")

        private fun skipWs(includeNewlines: Boolean) {
            while (i < n) {
                val c = text[i]
                if (c == ' ' || c == '\t' || c == '\r' || (includeNewlines && c == '\n')) i++
                else if (c == '#') {
                    while (i < n && text[i] != '\n') i++
                } else break
            }
        }

        private fun skipInlineWs() {
            while (i < n && (text[i] == ' ' || text[i] == '\t' || text[i] == '\r')) i++
        }

        private fun parseBasicString(): String {
            if (text[i] != '"') error("expected string")
            i++
            val sb = StringBuilder()
            while (true) {
                if (i >= n) error("unterminated string")
                when (val c = text[i++]) {
                    '"' -> return sb.toString()
                    '\\' -> {
                        if (i >= n) error("unterminated escape")
                        when (val e = text[i++]) {
                            '"' -> sb.append('"')
                            '\\' -> sb.append('\\')
                            'n' -> sb.append('\n')
                            't' -> sb.append('\t')
                            'r' -> sb.append('\r')
                            'b' -> sb.append('\b')
                            'u' -> {
                                if (i + 4 > n) error("bad \\u escape")
                                sb.append(text.substring(i, i + 4).toInt(16).toChar())
                                i += 4
                            }
                            'U' -> {
                                if (i + 8 > n) error("bad \\U escape")
                                sb.append(text.substring(i, i + 8).toInt(16).toChar())
                                i += 8
                            }
                            else -> error("unknown escape '\\$e'")
                        }
                    }
                    '\n' -> error("newline in string")
                    else -> sb.append(c)
                }
            }
        }

        private fun parseLiteralString(): String {
            if (text[i] != '\'') error("expected literal string")
            i++
            val sb = StringBuilder()
            while (true) {
                if (i >= n) error("unterminated string")
                val c = text[i++]
                if (c == '\'') return sb.toString()
                if (c == '\n') error("newline in string")
                sb.append(c)
            }
        }

        private fun parseKey(): String {
            skipInlineWs()
            return when {
                i < n && text[i] == '"' -> parseBasicString()
                i < n && text[i] == '\'' -> parseLiteralString()
                else -> {
                    val start = i
                    while (i < n && (text[i].isLetterOrDigit() || text[i] == '_' || text[i] == '-')) i++
                    if (i == start) error("expected key")
                    text.substring(start, i)
                }
            }
        }

        private fun parseArray(): List<Any?> {
            i++ // consume '['
            val list = mutableListOf<Any?>()
            while (true) {
                skipWs(includeNewlines = true)
                if (i < n && text[i] == ']') {
                    i++
                    return list
                }
                if (i >= n) error("unterminated array")
                list.add(parseValue())
                skipWs(includeNewlines = true)
                when {
                    i < n && text[i] == ',' -> i++
                    i < n && text[i] == ']' -> { i++; return list }
                    else -> error("expected ',' or ']' in array")
                }
            }
        }

        private fun parseInlineTable(): JsonObject {
            i++ // consume '{'
            val obj = JsonObject()
            skipWs(includeNewlines = false)
            if (i < n && text[i] == '}') {
                i++
                return obj
            }
            while (true) {
                skipWs(includeNewlines = false)
                val key = parseKey()
                skipInlineWs()
                if (i >= n || text[i] != '=') error("expected '=' in inline table")
                i++
                skipInlineWs()
                obj[key] = parseValue()
                skipWs(includeNewlines = false)
                when {
                    i < n && text[i] == ',' -> i++
                    i < n && text[i] == '}' -> { i++; return obj }
                    else -> error("expected ',' or '}' in inline table")
                }
            }
        }

        private fun parseValue(): Any? {
            skipInlineWs()
            if (i >= n) error("expected value")
            return when (val c = text[i]) {
                '"' -> parseBasicString()
                '\'' -> parseLiteralString()
                '[' -> parseArray()
                '{' -> parseInlineTable()
                't', 'f' -> {
                    if (text.startsWith("true", i)) { i += 4; true }
                    else if (text.startsWith("false", i)) { i += 5; false }
                    else error("invalid value")
                }
                else -> {
                    val start = i
                    while (i < n && text[i] != ',' && text[i] != ']' && text[i] != '}' &&
                        text[i] != '\n' && text[i] != '#'
                    ) i++
                    val token = text.substring(start, i).trim()
                    if (token.isEmpty()) error("expected value")
                    token.replace("_", "").toLongOrNull() ?: token
                }
            }
        }

        fun parse(): JsonObject {
            while (i < n) {
                skipWs(includeNewlines = true)
                if (i >= n) break
                if (text[i] == '[') {
                    // Table header (standard tables only).
                    i++
                    if (i < n && text[i] == '[') error("array-of-tables headers are not supported")
                    val path = mutableListOf<String>()
                    while (true) {
                        skipWs(includeNewlines = false)
                        path.add(parseKey())
                        skipWs(includeNewlines = false)
                        when {
                            i < n && text[i] == '.' -> i++
                            i < n && text[i] == ']' -> { i++; break }
                            else -> error("expected '.' or ']' in table header")
                        }
                    }
                    current = root
                    for (part in path) {
                        @Suppress("UNCHECKED_CAST")
                        current = current.getOrPut(part) { JsonObject() } as? JsonObject
                            ?: error("key '$part' is not a table")
                    }
                } else {
                    val key = parseKey()
                    skipInlineWs()
                    if (i >= n || text[i] != '=') error("expected '=' after key '$key'")
                    i++
                    val value = parseValue()
                    current[key] = value
                }
            }
            return root
        }
    }
}
